nixos/modules/installer/kexec/kexec-boot.nix

```diff
@@ -0,0 +1,50 @@
+# This module exposes a config.system.build.kexecBoot attribute,
+# which returns a directory with kernel, initrd and a shell script
+# running the necessary kexec commands.
+
+# It's meant to be scp'ed to a machine with working ssh and kexec binary
+# installed.
+
+# This is useful for (cloud) providers where you can't boot a custom image, but
+# get some Debian or Ubuntu installation.
+
+{ pkgs
+, modulesPath
+, config
+, ...
+}:
+{
+  imports = [
+    (modulesPath + "/installer/netboot/netboot-minimal.nix")
```

Comment:
Just for reference. We could a similar approach like in build-vm: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/build-vm.nix#L12
to make the system.build.kexecBoot available by default without having to import a profile.

Reply:
This imports a configuration.nix (`qemu-vm.nix`) as a "base config", while we just use `config.system.build.toplevel` to have this compose with the current config. Can we pass a `config.system.build` to `extendModules`? (cc @roberth)

Reply:
`extendModules` creates a clone of the current configuration, similar to `specialisations` with `inheritParentConfig = true`, except the clone isn't automatically wired into anything; it's just a value.

`nixos-rebuild build-vm` needs this because you're not booting your regular config, but rather an adaptation of it.

In the case of `kexecBoot`, this looks like a way of booting what you intend to be the regular config.
I don't see what would be the benefit of cloning the entire config and what changes you would make to it. It does have a performance cost, evaluating all modules twice.

Reply:
However we don't want to include `(modulesPath + "/installer/netboot/netboot-minimal.nix")` by default but we want to provide `system.build.kexecBoot` by default.

Reply:
> we don't want to include `(modulesPath + "/installer/netboot/netboot-minimal.nix")` by default

That's right. We can't, because it's essentially a profile module outside the `profiles` directory. Profile-ness spreads to anything that imports it, as such modules contain not just options but configs that are enabled by default.


> but we want to provide `system.build.kexecBoot` by default.

If you can write the important parts of the profile as a module, you can have `system.build.kexecBoot` without `imports`-ing anything, and only a very minor, generally accepted performance impact.

Do you have a real world use case for generating two very similar configurations where you are both netbooting and booting in some other way, with minimal differences between the two?

Reply:
I think it's probably easier to keep this a separately-to-be-imported module for now, instead of always exposing a `kexecBoot` attribute. It's pretty consistent with what else lives in these folders.

Reply:
Ok. Eventually I might add `nixos-rebuild kexec`, if I get to it (tm) :)

Reply:
Unresolved for visibility. This is discussion is still relevant.

Reply:
I've been somewhat critical of this because of eval performance, but the simplicity of such an interface is very much worth it for use cases where the kexec variant isn't evaluated often. https://github.com/NixOS/rfcs/pull/22 would make it a no-brainer.

I think we could add such an interface by reusing applying the both modules in the kexec variant.
 - kexec-only use cases can keep using the current method
 - use cases where the kexec variant is only evaluated once benefit from the improved UX
 - use cases where both the kexec variant and the regular config are always needed simultaneously in the same evaluator invocation also benefit from the improved UX
 - the documentation can be simplified. Improving the performance of kexec-only invocations can be a side note with 1 example that most users can skip